developers: Doc how to build against external PMIx/PRTE

jsquyres opened the pull request:
This has been on my to-do list for a while.

Reviewers: you can read the rendered version of this PR here: https://ompi--12946.org.readthedocs.build/en/12946/developers/building-open-mpi.html#building-against-external-openpmix-prrte

jsquyres committed:
developers: Doc how to build against external PMIx/PRTE

Signed-off-by: Jeff Squyres <jeff@squyres.com>

rhc54 commented on docs/developers/building-open-mpi.rst:
```diff
@@ -1,9 +1,87 @@
+     MPI this way, you may need to install the package manager's
+     "developer" Hwloc, Libevent, OpenPMIx, and/or PRRTE packages.
+
+1. Open MPI and PRRTE must be built against the **same** installation
```
Actually, not true - PMIx is designed to handle cross-version messaging. Only issue would be ensuring that the PMIx being used by PRRTE is new enough to support any OMPI-used features. Given the minimums we set, that shouldn't be a problem.

jsquyres replied:
Ok, fair point here.  I was really thinking about Hwloc and Libevent when I was writing this bullet.  I'll update.

rhc54 commented on docs/developers/building-open-mpi.rst:
```diff
@@ -1,9 +1,87 @@
+
+        Open MPI, OpenPMIx, and PRRTE must all use the same Hwloc and
+        Libevent libraries at run time (e.g., they must all resolve to
+        the same run-time loadable libraries at run time).
```
Not totally correct - PRRTE doesn't have to use the same as OMPI never loads "libprrte", so there is no potential for confusion.

jsquyres replied:
By transitive property, though, isn't this true?  OMPI must use the same hwloc + libevent as PMIX, and PRTE must use the same hwloc + libevent as PMIx, so therefore don't they all have to be the same?

bwbarrett replied:
you can have two versions of PMIx installed.  I don't think this is super common for a developer, but given how Slurm is packaged, may be more common there.

rhc54 commented on docs/developers/building-open-mpi.rst:
```diff
@@ -1,9 +1,87 @@
+
+1. Open MPI, OpenPMIx, and PRRTE must all be built against the
+   **same** installation of Hwloc and Libevent.  Meaning:
+
```
Not fully correct - PRRTE doesn't need to have the same.

jsquyres replied:
You can (will) run into super weird errors if PMIx, PRTE, and/or Open MPI are using different libevents and/or hwlocs at run-time.  They *should* be completely orthogonal, but we have definitely seen cases where they are not.  I'm ok documenting it as a **must** even if there are some cases where it actually does work to -- for example -- have PMIx run-time link against hwloc version X and PRTE run-time link against hwloc version Y.

bwbarrett replied:
Yeah, so something like:

1. Open MPI and the OpenPMIx library Open MPI links against must be built against the same installation of hwloc and Libevent.  It is not required that the runtime and Open MPI be built with the same version of PMIx, but the same hwloc/libevent linking rules also apply to PRRTE and its OPenPMIx library.

rhc54 commented on docs/developers/building-open-mpi.rst:
```diff
@@ -1,9 +1,87 @@
+
+Building Against External OpenPMIx / PRRTE
+------------------------------------------
+
```
I was under the impression that you were going to remove the ability to use an external PRRTE - no longer true? People can, of course, always use an independent version anyway, but this really referenced the ability to "link" `mpirun` to execute `prte` from the external PRRTE installation.

jsquyres replied:
Probably true.  But that's in the future -- this document is for `main` and `v5.0.x`, where you can still use an external PRTE.

rhc54 commented on docs/developers/building-open-mpi.rst:
```diff
@@ -1,9 +1,87 @@
+     manager.  Assuming that the package-manager installs of OpenPMIx
+     and PRRTE were built against the package-manager-provider Hwloc
+     and Libevent, then Open MPI will *also* need to be built against
+     the package-manager-provided Hwloc and Libevent.  To build Open
```
Again, true for PMIx but not for PRRTE

rhc54 reviewed (dismissed):
I think there is some confusion about PRRTE - see comments.

rhc54 commented:
I think it stalled out. The problem is that the combinations are a little complex because PRRTE is totally independent of OMPI. So you really cannot make statements about requirements that link the two. What I'd suggest is to not try to make such sweeping statements and instead say what OMPI needs, and then say what PRRTE needs.

For OMPI, you need PMIx to use the same hwloc and libevent libs that OMPI is using because they share the same memory space.

For PRRTE, you need PMIx to use the same hwloc and lbievent libs that PRRTE is using because they share the same memory space. There is _no_ requirement that these be the same that OMPI is using because OMPI does not link against PRRTE.

_**IF**_ OMPI and PRRTE are using the _same_ PMIx, then that means that all three will wind up using the same hwloc and libevent. _However_, there is _no_ requirement that PRRTE and OMPI use the _same_ PMIx installation. PMIx supports cross-version operations, and so the two can not only use different installations, but the two installations don't even need to be the same PMIx version.

jsquyres committed:
TO BE SQUASHED

Separated out into a separate commit just for reviewing purposes.

Signed-off-by: Jeff Squyres <jeff@squyres.com>

jsquyres commented:
@rhc54 I just rebased and updated these docs per the unresolved comments from this stale PR.  Can you re-review?  I have the changes from your comments in a 2nd commit (to be squashed when we're done) so that it's easy to see what I changed.
